File: src/find_ids.py
"""
find_ids.py — Utility to discover Slack channel IDs and ClickUp workspace structure.
Run this whenever you need to look up IDs for configuring the weekly-digest pipeline.
Requires SLACK_BOT_TOKEN and CLICKUP_API_TOKEN in a .env file.
"""

import requests
import os
from dotenv import load_dotenv

load_dotenv()

# ──────────────────────────────────────────────
# Constants
# ──────────────────────────────────────────────
SLACK_API   = "https://slack.com/api"
CLICKUP_API = "https://api.clickup.com/api/v2"

SLACK_PAGE_LIMIT = 200  # max channels per page (Slack caps at 200)

# ──────────────────────────────────────────────
# Slack setup
# Bot token scopes needed:
#   • channels:read  → list public channels
#   • groups:read    → list private channels the bot has been invited to
# ──────────────────────────────────────────────
headers_slack = {"Authorization": f"Bearer {os.getenv('SLACK_BOT_TOKEN')}"}


def get_all_slack_channels():
    """Fetch ALL visible channels using cursor-based pagination."""
    channels = []
    cursor = None
    channel_types = "public_channel,private_channel"

    while True:
        params = {
            "limit": SLACK_PAGE_LIMIT,
            "types": channel_types,
        }
        if cursor:
            params["cursor"] = cursor

        r = requests.get(
            f"{SLACK_API}/conversations.list",
            headers=headers_slack,
            params=params,
        )
        data = r.json()

        if not data.get("ok"):
            error = data.get("error")
            print(f"   Slack API error: {error}")
            break

        channels.extend(data.get("channels", []))

        # Cursor-based pagination: keep going while there's a next_cursor
        next_cursor = data.get("response_metadata", {}).get("next_cursor")
        if not next_cursor:
            break
        cursor = next_cursor

    return channels


# ──────────────────────────────────────────────
# Print Slack channels
# ──────────────────────────────────────────────
print("\n=== SLACK CHANNELS (all, paginated) ===")
all_channels = get_all_slack_channels()

for c in sorted(all_channels, key=lambda x: x["name"]):
    visibility = "private" if c.get("is_private") else "public"
    members = c.get("num_members", "?")
    print(f"  #{c['name']:35s} → ID: {c['id']}  ({visibility}, {members} members)")

print(f"\nTotal: {len(all_channels)} channels found")

# ──────────────────────────────────────────────
# ClickUp setup
# Needs a personal API token (v2) set as CLICKUP_API_TOKEN in .env
# ──────────────────────────────────────────────
headers_cu = {"Authorization": os.getenv("CLICKUP_API_TOKEN")}

print("\n=== CLICKUP STRUCTURE ===")

# 1. Get workspace (team)
teams = requests.get(f"{CLICKUP_API}/team", headers=headers_cu).json()
team_id = teams["teams"][0]["id"]
print(f"Team: {teams['teams'][0]['name']} (ID: {team_id})")

# 2. List spaces inside the workspace
spaces = requests.get(
    f"{CLICKUP_API}/team/{team_id}/space?archived=false",
    headers=headers_cu,
).json()

# 3. Walk spaces → folders → lists and print the full tree
for space in spaces.get("spaces", []):
    print(f"\n  Space: {space['name']} → ID: {space['id']}")

    folders = requests.get(
        f"{CLICKUP_API}/space/{space['id']}/folder?archived=false",
        headers=headers_cu,
    ).json()

    for folder in folders.get("folders", []):
        print(f"    Folder: {folder['name']:40s} → ID: {folder['id']}")

        lists = requests.get(
            f"{CLICKUP_API}/folder/{folder['id']}/list?archived=false",
            headers=headers_cu,
        ).json()

        for lst in lists.get("lists", []):
            print(f"      List: {lst['name']:35s} → ID: {lst['id']}  ({lst.get('task_count', '?')} tasks)")

File: src/find-ids.ts
/**
 * find-ids.ts — Utility to discover Slack channel IDs and ClickUp workspace structure.
 * Run this whenever you need to look up IDs for configuring the weekly-digest pipeline.
 * Requires SLACK_BOT_TOKEN and CLICKUP_API_TOKEN in a .env file.
 *
 * Usage: pnpm find-ids
 */

import "dotenv/config";
import logger from "./config/logger.js";

// Constants

const SLACK_API = "https://slack.com/api";
const CLICKUP_API = "https://api.clickup.com/api/v2";
const SLACK_PAGE_LIMIT = 200;

// Setup

const slackHeaders: HeadersInit = {
    Authorization: `Bearer ${process.env.SLACK_BOT_TOKEN}`,
};

const clickupHeaders: HeadersInit = {
    Authorization: process.env.CLICKUP_API_TOKEN ?? "",
};

interface SlackChannel {
    id: string;
    name: string;
    is_private: boolean;
    num_members: number;
}

async function getAllSlackChannels(): Promise<SlackChannel[]> {
    const channels: SlackChannel[] = [];
    let cursor: string | undefined;

    while (true) {
        const params = new URLSearchParams({
            limit: String(SLACK_PAGE_LIMIT),
            types: "public_channel,private_channel",
        });
        if (cursor) params.set("cursor", cursor);

        const res = await fetch(`${SLACK_API}/conversations.list?${params}`, {
            headers: slackHeaders,
        });
        const data = await res.json();

        if (!data.ok) {
            logger.error({ error: data.error }, "Slack API error");
            break;
        }

        channels.push(...(data.channels ?? []));

        const nextCursor = data.response_metadata?.next_cursor;
        if (!nextCursor) break;
        cursor = nextCursor;
    }

    return channels;
}

// Compact one-line helpers (avoid pino's multi-line structured output for large lists)
const pad = (s: string, n: number) => s.padEnd(n);

// Runner

async function run() {
    // ── Slack channels ───────────────────────────────────────────────────
    console.log("\n=== SLACK CHANNELS ===\n");
    const allChannels = await getAllSlackChannels();
    const sorted = allChannels.sort((a, b) => a.name.localeCompare(b.name));

    for (const c of sorted) {
        const vis = c.is_private ? "priv" : "pub ";
        console.log(
            `  #${pad(c.name, 35)} ${pad(c.id, 14)} ${vis}  ${c.num_members ?? "?"} members`
        );
    }
    console.log(`\n  Total: ${allChannels.length} channels\n`);

    // ── Slack users ─────────────────────────────────────────────────────
    console.log("=== SLACK USERS ===\n");
    const users: Array<{ id: string; name: string; real_name: string; is_bot: boolean }> = [];
    let userCursor: string | undefined;

    while (true) {
        const params = new URLSearchParams({ limit: "200" });
        if (userCursor) params.set("cursor", userCursor);

        const res = await fetch(`${SLACK_API}/users.list?${params}`, {
            headers: slackHeaders,
        });
        const data = await res.json();

        if (!data.ok) {
            if (data.error === "missing_scope") {
                console.log(
                    "  ⚠ Missing scope: users:read — add it in your Slack app settings and reinstall\n"
                );
            } else {
                logger.error({ error: data.error }, "Failed to list Slack users");
            }
            break;
        }

        users.push(...(data.members ?? []));

        const next = data.response_metadata?.next_cursor;
        if (!next) break;
        userCursor = next;
    }

    if (users.length > 0) {
        const people = users
            .filter((u) => !u.is_bot && u.id !== "USLACKBOT")
            .sort((a, b) => (a.real_name ?? a.name).localeCompare(b.real_name ?? b.name));

        for (const u of people) {
            console.log(`  ${pad(u.real_name || u.name, 35)} ${u.id}`);
        }
        console.log(`\n  Total: ${people.length} users\n`);
    }

    // ── ClickUp structure ────────────────────────────────────────────────
    console.log("=== CLICKUP STRUCTURE ===\n");

    const teamsRes = await fetch(`${CLICKUP_API}/team`, { headers: clickupHeaders });
    const teams = await teamsRes.json();
    const teamId = teams.teams[0].id;
    console.log(`  Team: ${teams.teams[0].name} (${teamId})\n`);

    const spacesRes = await fetch(`${CLICKUP_API}/team/${teamId}/space?archived=false`, {
        headers: clickupHeaders,
    });
    const spaces = await spacesRes.json();

    for (const space of spaces.spaces ?? []) {
        console.log(`  Space: ${space.name}  →  ${space.id}`);

        const foldersRes = await fetch(`${CLICKUP_API}/space/${space.id}/folder?archived=false`, {
            headers: clickupHeaders,
        });
        const foldersData = await foldersRes.json();

        for (const folder of foldersData.folders ?? []) {
            console.log(`    Folder: ${pad(folder.name, 40)}  →  ${folder.id}`);

            const listsRes = await fetch(`${CLICKUP_API}/folder/${folder.id}/list?archived=false`, {
                headers: clickupHeaders,
            });
            const listsData = await listsRes.json();

            for (const lst of listsData.lists ?? []) {
                console.log(
                    `      List: ${pad(lst.name, 35)}  →  ${pad(lst.id, 14)}  (${lst.task_count ?? "?"} tasks)`
                );
            }
        }
        console.log();
    }
}

run().catch((e) => {
    logger.fatal({ err: e }, "Unhandled error");
    process.exit(1);
});
